core/src/dpdk/stats.rs

```diff
@@ -0,0 +1,185 @@
+use super::{CoreId, Port, PortId};
+use crate::ffi::{self, ToResult};
+use crate::metrics::{labels, Key, Measurement};
+use crate::Result;
+
+/// Port stats collector.
+pub struct PortStats {
+    id: PortId,
+    name: String,
+    cores: Vec<CoreId>,
+}
+
+impl PortStats {
+    /// Builds a collector from the port.
+    pub fn build(port: &Port) -> Self {
+        // sorts the core ids, so they are in the same order as the stats.
+        let mut cores = port.queues().keys().copied().collect::<Vec<_>>();
+        cores.sort();
+
+        PortStats {
+            id: port.id(),
+            name: port.name().to_owned(),
+            cores,
+        }
+    }
+
+    /// Returns the port name.
+    pub fn name(&self) -> &str {
+        self.name.as_str()
+    }
+
+    /// Collects the port stats tracked by DPDK.
+    pub fn collect(&self) -> Result<Vec<(Key, Measurement)>> {
+        let mut stats = ffi::rte_eth_stats::default();
+        unsafe {
+            ffi::rte_eth_stats_get(self.id.raw(), &mut stats).to_result()?;
+        }
+
+        let mut values = Vec::new();
+
+        if ffi::RTE_ETHDEV_QUEUE_STAT_CNTRS >= self.cores.len() as u32 {
+            // packet and byte counts are being tracked per core.
+            self.cores.iter().enumerate().for_each(|(idx, core_id)| {
+                values.push((
```

Comment:
since this gets called over and over again w/ a mostly familiar structure, can this push be a function where labels are passed in or where you have some set ones (for dir/port vs dir/port/core)?

Reply:
ok addressed. definitely way cleaner now.